File: go/linkedlist/leetcode/reverse_linked_list.go
package main

//反转链表

//迭代
func reverseList1(head *ListNode) *ListNode {
	var prev *ListNode
	for head != nil {
		head.Next, prev, head = prev, head, head.Next
	}
	return prev
}

//递归
func reverseList(head *ListNode) *ListNode {
	return helper(nil, head)
}

func helper(prev, cur *ListNode) *ListNode {
	if cur == nil {
		return prev
	}
	next := cur.Next
	cur.Next = prev
	return helper(cur, next)
}

//官方的递归解法
func reverseList2(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return head
	}

	p := reverseList2(head.Next)
	head.Next.Next = head
	head.Next = nil
	return p
}

File: python/linkedlist/leetcode/reverse_linked_list.py
from typing import List


class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:

    # 迭代
    def reverseList1(self, head: ListNode) -> ListNode:
        prev, back = None, head
        while back:
            back.next, prev, back = prev, back, back.next
        return prev

    # 递归
    def reverseList2(self, head: ListNode) -> ListNode:
        def helper(prev, cur) -> ListNode:
            if not cur:
                return prev
            n = cur.next
            cur.next = prev
            return helper(cur, n)
        return helper(None, head)
        